docs/social/online-matches/index.md

```diff
@@ -0,0 +1,266 @@
+so maybe a graph database is the right fit? Well, Kafka Connect makes
+it easy to use the strengths of both in tandem, but before we go that
+far, let's dig in and see how much graph-analysis we can do with some
+clever ksqlDB tricks._
+
+We'll create a messaging system, and then ask, "Which pairs of people
+have made proper connections?" Defining a 'proper' connection is a
+mystery for sociologists, but for our purposes we'll say two people
+have connected properly if A sends a message to B, B responds, and A
+follows-up. That should be enough chit-chat to get the ball rolling.
+
+# Solution
+
+``` sql
+--8<-- "docs/social/online-matches/process.sql"
+```
+
+## Explanation
+
+### Set Up Your Environment
+
+--8<-- "docs/shared/ccloud_setup.md"
+
+### Create The Initial Dataset
+
+``` sql
+--8<-- "docs/social/online-matches/manual.sql"
+```
+
+## Tracking Connections
+
+If you look at those `messages`, it's clear there are a lot of hellos
+bouncing around, but beyond that it's hard to see any patterns. Let's
+use some queries to make sense of it. We'll build up our answer to,
+"Who's connected to who?"  gradually. Before we begin, some session
+settings to make sure we all get the same results:
+
+```sql
+SET 'auto.offset.reset' = 'earliest';
+SET 'ksql.query.pull.table.scan.enabled' = 'true';
+```
+
+### Split By Conversation
+
+The first step is to break the stream up into individual
+conversations. If we sort the sender and receiver of each message, we
+can create a unique ID for every pair that chats (or tries to start
+chatting), and use that to group all the events:
+
+```sql
+CREATE TABLE conversations_v1 AS
+  SELECT
+    ARRAY_JOIN(ARRAY_SORT(ARRAY [send_id, recv_id]), '<>') AS conversation_id,
+    COLLECT_LIST(rowtime) AS message_times
+  FROM messages
+  GROUP BY 
+    ARRAY_JOIN(ARRAY_SORT(ARRAY [send_id, recv_id]), '<>');
+```
+
+Querying that:
+
+```sql
+SELECT * FROM conversations_v1;
+```
+
+```txt
++----------------+----------------------------------------------+
+|CONVERSATION_ID |MESSAGE_TIMES                                 |
++----------------+----------------------------------------------+
+|3<>4            |[1637318173288, 1637317383974, 1637318170245] |
+|4<>5            |[1637317384066, 1637317384210, 1637317384126] |
+|1<>2            |[1637317383692, 1637317383832, 1637317383887] |
+|1<>3            |[1637317383778]                               |
+```
+
+(_Note that because we sorted the `[send_id, recv_id]` array, it doesn't matter if 1
+was sending to 2 or 2 was sending to 1 - we get the same conversation ID for both
+directions._)
+
+### Chat By Chat
+
+That's a big help - we can analyse conversations individually. 
+
+
+Let's put each of those conversations in time order, and keep track of the sender as we go.
+
+We do this in two steps - we'll enhance our `message_times` column to
+build up a map with the `rowtime` as the key, and the `send_id` as the
+value:
+
+```sql
+CREATE TABLE conversations_v2 AS
```

Comment:
v2 seems like the same as v1, plus send_id.  What do you think about merging these?

Reply:
Well, the big idea that v2 introduces is collecting two lists together and using them as the key-value pairs for a map. That seemed like quite a large jump that was worth outlining separately. Certainly when I was deciphering Nick's query, I was considering it as a large step.